### DISCUSSION TITLE
A sampling method of production costs may be wrong

### DISCUSSION THREAD

**Discussion by a4rcvv** [yasserfarouk/scml#99] [one-shot-qa] [answered]
According to the [description for OneShot track](http://www.yasserm.com/scml/scml2022oneshot.pdf), a production cost $m_a$ is sampled as follows:

1. Let $m_l$ be $l \times U[1,10]$, where $l$ is a level of an agent.
2. $m_a$ is sampled from $U[m_l, 4 \times m_l]$.

However, in fact, it seems that $m_a$ is sampled as follows:

1. $m$ is sampled from a discrete uniform distribution $U[1,4]$.
1. Let $m_a$ be $ l \times m$, where $l$ is the level of an agent.

Is the description wrong, or is this unintended behavior?

Here is a code for verification:

```python
from scml.oneshot import RandomOneShotAgent, SCML2020OneShotWorld

if __name__ == '__main__':
    seller_cost_count = [0] * 9
    buyer_cost_count = [0] * 9

    for i in range(1000):
        world = SCML2020OneShotWorld(
            **SCML2020OneShotWorld.generate(
                [RandomOneShotAgent, RandomOneShotAgent],
                agent_processes=[0, 1],
                random_agent_types=False,
                n_steps=3,
            ),
        )

        seller_id = list(world.agent_profiles.keys())[0]
        seller_cost = world.agent_profiles[seller_id].cost
        assert seller_cost.is_integer()
        seller_cost = int(seller_cost)
        seller_cost_count[seller_cost] += 1

        buyer_id = list(world.agent_profiles.keys())[1]
        buyer_cost = world.agent_profiles[buyer_id].cost
        assert buyer_cost.is_integer()
        buyer_cost = int(buyer_cost)
        buyer_cost_count[buyer_cost] += 1

    print(seller_cost_count)
    print(buyer_cost_count)
```

And this is the output of it.

```
[0, 236, 261, 241, 262, 0, 0, 0, 0]
[0, 0, 245, 0, 240, 0, 240, 0, 275]
```

**Comment by yasserfarouk** [comment, marked as answer]
The behavior in the code is the intended behavior and the description should have been updated.

The behavior in the description was the initial design but we noticed that it makes the scores too much dependent on the production cost so we switched to the second approach which has smaller variation in production cost per level. I forgot to update the description. Will do that soon and notify all participants.

Please note that agent’s behavior should not depend much on these distributions as they are likely to change in the future (I.e. 2023). The agent can read its own production cost directly from the AWI and should minimize its dependence on the exact limits of the distribution used for sampling.

**Reply by a4rcvv** [reply]
Thank you for the clarification!
